src/iop/lut3d.c

```diff
@@ -0,0 +1,1023 @@
+#include "develop/imageop.h"
+#include "dtgtk/button.h"
+#include "gui/gtk.h"
+#include "iop/iop_api.h"
+
+#include <gtk/gtk.h>
+#include <libgen.h>
+#include <png.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#if defined (_WIN32)
+#include "win/getdelim.h"
+#endif // defined (_WIN32)
+
+DT_MODULE_INTROSPECTION(1, dt_iop_lut3d_params_t)
+
+typedef enum dt_iop_lut3d_colorspace_t
+{
+  DT_IOP_SRGB = 0,
+  DT_IOP_REC709 = 1,
+  DT_IOP_LIN_REC709 = 2,
+  DT_IOP_LIN_PROPHOTORGB = 3,
+} dt_iop_lut3d_colorspace_t;
+
+typedef enum dt_iop_lut3d_interpolation_t
+{
+  DT_IOP_TETRAHEDRAL = 0,
+  DT_IOP_TRILINEAR = 1,
+  DT_IOP_PYRAMID = 2,
+} dt_iop_lut3d_interpolation_t;
+
+typedef void((*dt_interpolation_worker)(float *input, float *output, const float *const clut, const uint8_t level));
+
+typedef struct dt_iop_lut3d_params_t
+{
+  char filepath[512];
+  int colorspace;
+  int interpolation;
+} dt_iop_lut3d_params_t;
+
+typedef struct dt_iop_lut3d_gui_data_t
+{
+//  GtkWidget *lutfolder;
+  GtkWidget *hbox;
+  GtkWidget *filepath;
+  GtkWidget *colorspace;
+  GtkWidget *interpolation;
+} dt_iop_lut3d_gui_data_t;
+
+typedef struct dt_iop_lut3d_data_t
+{
+  dt_iop_lut3d_params_t params;
+  float *clut;  // cube lut pointer
+  uint8_t level; // cube_size
+} dt_iop_lut3d_data_t;
+
+typedef struct dt_iop_lut3d_global_data_t
+{
+  int kernel_lut3d_tetrahedral;
+  int kernel_lut3d_trilinear;
+  int kernel_lut3d_pyramid;
+} dt_iop_lut3d_global_data_t;
+
+const char *name()
+{
+  return _("lut 3D");
+}
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_SUPPORTS_BLENDING;
+}
+
+int groups()
+{
+  return IOP_GROUP_COLOR;
+}
+
+int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  return iop_cs_rgb;
+}
+
+// From `HaldCLUT_correct.c' by Eskil Steenberg (http://www.quelsolaar.com) (BSD licensed)
+void correct_pixel_trilinear(float *input, float *output, const float *const clut, const uint8_t level)
+{
+  int red, green, blue, i, j;
+  float tmp[6];
+  const int level2 = level * level;
+
+  for(int c = 0; c < 3; ++c) input[c] = input[c] < 0.0f ? 0.0f : (input[c] > 1.0f ? 1.0f : input[c]);
+  red = input[0] * (float)(level - 1);
+  if(red > level - 2) red = level - 2;
+  if(red < 0) red = 0;
+  green = input[1] * (float)(level - 1);
+  if(green > level - 2) green = level - 2;
+  if(green < 0) green = 0;
+  blue = input[2] * (float)(level - 1);
+  if(blue > level - 2) blue = level - 2;
+  if(blue < 0) blue = 0;
+
+  const float r = input[0] * (float)(level - 1) - red; // delta red
+  const float g = input[1] * (float)(level - 1) - green; // delta green
+  const float b = input[2] * (float)(level - 1) - blue; // delta blue
+
+// indexes of P000 to P111 in clut
+  const int color = red + green * level + blue * level * level;
+  i = color * 3;  // P000
+  j = (color + 1) * 3;  // P100
+
+  tmp[0] = clut[i++] * (1 - r) + clut[j++] * r;

```

Comment:
Please use `clut[i]`, then `clut[i+1]` and finally `clut[i+2]` instead of `clut[i++]` so, again, the compiler will understand with no ambiguity that we work on contiguous data (again, for auto SIMD). Also, it makes the code more readable.

Reply:
done